### ISSUE TITLE
Defining relative paths in the project json

### ISSUE THREAD

**Issue by JoelRV** [qupath/qupath#346]
Hi Peter,
I was trying to make a project folder sharable by sending the image files in an annexing folder. My Idea was initially to modify the project json image paths manualy to either make them relative to the project dir using "../image/image.file" or "{$PROJECT_DIR}/image/image.file".

It seems that since the milestone releases the json structure and image loading changed substantially. Thus when I try the above solution, I get:

```
"uri": "file:/../image/image1.tif" >>>  I get the new prompt window to change the URIs

"uri": "file:/{$PROJECT_DIR}/../image/image1.tif" >>> Error : java.net.URISyntaxException: Illegal character in path at index 6:....
```

So my guess is simply that the new uri process broke those old workarounds. Is the an other way to either use the "../" syntax or to use a project dir variable?

I know that #325 specifies that the m3 version should recognize relative paths, but I simply can't get it to work.

Desktop (please complete the following information):

    OS: Linux Ubuntu 16.04
    QuPath Version 0.2.0-m3 vs 0.1.2

**Comment by petebankhead** [comment]
Have you tried creating a new project with m3 and seeing how it behaves?

m3 avoids the use of `{$PROJECT_DIR}` as a QuPath-specific solution is rather hard to maintain, and switches to URIs to overcome backslash/forward slash issues and also support OMERO-hosted images as well as local files.

Rather, what m3 should do is:
* Always store the absolute URI for the image
* Store the last known URI for any project file
* Check for the existence of the absolute URI
  * If that succeeds, good!
  * If that fails, relativize the URI based on the last known project file URI and try it
* Show a dialog box to confirm any required URI changes / missing URIs requiring manual correction

**Comment by JoelRV** [comment]
Hi,
Thanks for the reply and sorry for late feedback.

So, I tested it all with a fresh install of m3 and a clean dedicated folder for testing. the structure is as follows:

-TestFolder
-- ImageFolder
-- qupath_project_folder1
-- qupath_project_folder2

For both projects, I generate a fresh initialization, load all the images in the ImageFolder and do some operations (shapes, cell detection ...).

My aim is to be able to open either of both projects and it finding the relevant image files relatively (going one folder up), especially after moving the TestFolder around.

Now I tested it on different PCs and VMs and keep getting the same behavior, i.e. getting the prompt window to manually re-specify the URI paths. So either I am missing something or I am going wrong about it.

While the prompt is not a problem for me per se, when sending the files to collaborators, people tend to get scared or confused by the prompt, so I wanted to manually specify relative paths to make it more user friendly.

**Comment by petebankhead** [comment]
Ah ok, the prompt _should_ appear but it should be pre-populated with the correct URIs. However agreeing to the changes will result in the absolute paths being overwritten.

Consequently, one might wish to copy the '.qpproj' file first if it's desirable to have a project file with both sets of absolute URIs - updating one and leaving the other untouched. This does however become more complicated if images are later added or removed for one project.

This is the intended behavior, although would be good to discuss if there is a better choice.

My concern with depending upon relative paths _only_ is that if the project moves - but images remain in the same place - then the project will also break. It becomes essential to move both together.

Also, if the absolute paths are _not_ updated whenever relative paths 'work' then this means that QuPath needs to silently make adjustments to URIs - and _potentially_ could end up selecting a wrong image that happens to have the same name.

**Comment by JoelRV** [comment]
Indeed, the prompt does not prepopulate new suggested paths in my case.

While I don't see an easy fit-all solution, in my case the simple fact to be able to specify the path based on the current location of the project folder (for example storing it in a variable at the moment of opening) would already solve the problem (like it used to in v0.2.0 using $PROJECT_DIR).

This means that before sharing the project I could manually modify the JSON if needed. 

My ideal scenario would be :
1) Test specified path using current project folder absolute path as a starting point (using a $PROJECT_DIR)
2) Test relative path if 1) fails
3) Make user manually specify URI if all fails

This being said, It might not be as straightforward to implement code-wise depending on how you are currently rehauling your code base and the fact that there might be other fires to tackle first.

**Comment by petebankhead** [comment]
Hmmm, if I set up a similar folder structure to what you describe, then rename the base directory, this is what I see:
![image](https://user-images.githubusercontent.com/4690904/62463695-91b99180-b782-11e9-85dd-5eb2e17dbdff.png)
The paths on the right indicate the 'fixed' relative paths and if I choose 'Apply changes' then everything works.

Can you see what is different in your case?

I think what you're describing is similar to what actually happens (or should happen), except that a prompt is shown if 2. is reached. Potentially this prompt could be made less obtrusive if QuPath can identify that relativizing works for all URIs and there is nothing to manually specify.

**Comment by JoelRV** [comment]
Well I do rename my test directory from PDA_SHOWCASE to PDA_SHOWCASE2, without changing anything, I get this:
![image](https://user-images.githubusercontent.com/18205832/62475608-a52a2400-b7a5-11e9-89e6-bde75f930c43.png)

As I see there is no replacement suggestion. Is this possibly linked with being on a Linux based system?

**Comment by petebankhead** [comment]
Yes! I think that's it, thanks. I won't be able to check Linux until next week, but I could reproduce the issue on my Mac. It seems that resolving relative paths using 'regular' files (as opposed to directories) was only working successfully on Windows.

Hopefully the above commit will fix the issue.

**Comment by petebankhead** [comment]
Closing this issue as I think it is addressed by v0.2.0-m4. Please reopen if not.
